### PR TITLE
Fix a GUS audio regression in Star Control II (#1530)

### PR DESCRIPTION
**Pull request by kcgen**
This PR fixes a regression in the GUS emulation where it could use incomplete-sounding voices in Star Control 2. 

It does some code duplication cleanup in the GUS and fixes a Staging bug where the GUS IRQ was limited to 9 (when it should allow up to 12). 

It uses [a community patch](https://www.vogons.org/viewtopic.php?t=28331) that implements contiguous DMA transfers similar to what EMS386.EXE does -- thanks to @peterferrie from the Vogons forum, and then refactors that along with more code de-duping in the DMA area. These are are stand-alone commits meant to be bisectable if there are problems.

DMA transfers were tested using Quake and SC2, and run with UBSAN and ASAN instrumentation.

https://user-images.githubusercontent.com/1557255/150623662-70e62668-2d37-4e07-aae3-932d25b026b7.mp4

https://user-images.githubusercontent.com/1557255/150623663-15131b29-5fee-41f6-b9cc-5c934be4f135.mp4

Thanks to @marsej for the report and detailed comparisons!

### CODE REVIEW

**Review by LowLevelMahn — COMMENTED** [review]
good style, ok from my side

**Comment by johnnovak** [comment]
@kcgen I saw you included me in the list of reviewers, had a brief look, but couldn't meaningfully review it as i'm unfamiliar with the GUS. Style looks great though, as always :)

**Comment by kcgen** [comment]
> @kcgen I saw you included me in the list of reviewers, had a brief look, but couldn't meaningfully review it as i'm unfamiliar with the GUS. Style looks great though, as always :)

@johnnovak , yup - just looking for general feedback, logic improvements, coding style/comments/variables names - the usual :-)  Thanks for the feedback!

**Comment by kcgen** [comment]
@marseg, thank you for the bug reportm and @johnnovak for the review. 

State of the PR:
 - Addressed the issues noted in the ticket.
 - Passing playback in 42 GUS-native games (plus modtracker, demo, and wavetable tests).
 - Passing UBSAN and ASAN runs.
 - Passing non-GUS related checks in MIDI, VESA, and Win31.
 - Commit names are OK.
 - Each commit individually compiles (important for future bisecting).
 
I've taken a final pass through the code and there are no further review comments (after sitting for 13 days), so I think it's time to merge.

File: src/hardware/gus.cpp [hunk 1]
```diff
@@ -715,17 +719,37 @@ void Gus::CheckVoiceIrq()
...
+	uint32_t adjusted;
+	if (IsDmaXfer16Bit()) {
+		const auto upper = offset & 0b1100'0000'0000'0000'0000;
+		const auto lower = offset & 0b0011'1111'1111'1111'1110;
+		adjusted = upper | (lower >> 1);
+	}
+	else {
+		adjusted = offset;
+	}
+	dma_addr = check_cast<uint16_t>(adjusted >> 4); // pack it into the 16-bit register
```
Comment (kklobe) [inline src/hardware/gus.cpp:751]: This line is causing Quake to assert with GUS enabled. If `adjusted` is a 24-bit value, `adjusted >> 4` will give a 20-bit value, causing the `check_cast` to fail. Do we want this to be either a `static_cast` or change it to e.g.
```cpp
	dma_addr = check_cast<uint16_t>(adjusted >> 4 & 0xffff);
```
?

Comment (kklobe) [reply src/hardware/gus.cpp:751]: Quake output:
```log
022-06-18 21:50:09.595 | GUS: DMA address adjusted >> 4 = fa00
2022-06-18 21:50:09.595 | GUS: DMA address adjusted >> 4 = fb00
2022-06-18 21:50:09.595 | GUS: DMA address adjusted >> 4 = fc00
2022-06-18 21:50:09.595 | GUS: DMA address adjusted >> 4 = fd00
2022-06-18 21:50:09.595 | GUS: DMA address adjusted >> 4 = fe00
2022-06-18 21:50:09.595 | GUS: DMA address adjusted >> 4 = ff00
2022-06-18 21:50:09.595 | GUS: DMA address adjusted >> 4 = 10000
Assertion failed: (static_cast<next_int_t<check_t>>(in) <= static_cast<next_int_t<cast_t>>(std::numeric_limits<cast_t>::max())), function check_cast, file support.h, line 208.
```

Comment (kcgen) [reply src/hardware/gus.cpp:751]: Thank you, @kklobe! Right; with exactly 1 MB of ram, that's 20 bits of addressable space  (2 ** 20 = 1048576), and given the first 4 bits are shifted off the end, that leaves the top 16bit in the 20 bit range. :+1:  Applied and regression tested.

**Pull request merged**